src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,728 @@
+			 src_fmt->frame_fmt);
+		return -EINVAL;
+	}
+
+	buffer_format = layout_convert_sof_to_me(component->ca_source->buffer_fmt);
+	if (buffer_format < 0) {
+		comp_err(dev, "waves_effect_init() sof buffer format %d not supported",
+			 component->ca_source->buffer_fmt);
+		return -EINVAL;
+	}
+
+	sample_bytes = sample_format_convert_to_bytes(sample_format);
+	if (sample_bytes < 0) {
+		comp_err(dev, "waves_effect_init() sample_format %d not supported",
+			 sample_format);
+		return -EINVAL;
+	}
+
+	waves_codec->request_max_bytes = 0;
+	waves_codec->response_max_bytes = 0;
+	waves_codec->response = 0;
+	waves_codec->i_buffer = 0;
+	waves_codec->o_buffer = 0;
+
+	waves_codec->i_format.sampleRate = src_fmt->rate;
+	waves_codec->i_format.numChannels = src_fmt->channels;
+	waves_codec->i_format.samplesFormat = sample_format;
+	waves_codec->i_format.samplesLayout = buffer_format;
+
+	waves_codec->o_format = waves_codec->i_format;
+
+	waves_codec->sample_size_in_bytes = sample_bytes;
+	waves_codec->buffer_samples = src_fmt->rate * 2 / 1000; /* 2 ms io buffers */
+	waves_codec->buffer_bytes = waves_codec->buffer_samples * src_fmt->channels *
+		waves_codec->sample_size_in_bytes;
+
+	comp_info(dev, "waves_effect_init() rate %d, channels %d", waves_codec->i_format.sampleRate,
+		  waves_codec->i_format.numChannels);
+
+	comp_info(dev, "waves_effect_init() format %d, layout %d, frame %d",
+		  waves_codec->i_format.samplesFormat, waves_codec->i_format.samplesLayout,
+		  waves_codec->buffer_samples);
+
+	status = MaxxEffect_Initialize(waves_codec->effect, i_formats, 1, o_formats, 1);
+
+	if (status) {
+		comp_err(dev, "waves_effect_init() MaxxEffect_Initialize() error %d",
+			 status);
+		return -EINVAL;
+	}
+
+	comp_dbg(dev, "waves_effect_init() done");
+	return 0;
+}
+
+/* allocate additional buffers for MaxxEffect */
+static int waves_effect_buffers(struct comp_dev *dev)
+{
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec = codec->private;
+	MaxxStatus_t status;
+
+	comp_dbg(dev, "waves_effect_buffers() start");
+
+	/* memmory for response
+	 * SOF does not support get requests from codec adapter right now
+	 * response will be stored in internal buffer to be dumped into logs
+	 */
+	status = MaxxEffect_GetMessageMaxSize(waves_codec->effect, &waves_codec->request_max_bytes,
+					      &waves_codec->response_max_bytes);
+
+	if (status) {
+		comp_err(dev, "waves_effect_buffers() MaxxEffect_GetMessageMaxSize() error %d",
+			 status);
+		return -EINVAL;
+	}
+
+	waves_codec->response = codec_allocate_memory(dev, waves_codec->response_max_bytes, 16);
+	if (!waves_codec->response) {
+		comp_err(dev, "waves_effect_buffers() memory for response");
+		return -ENOMEM;
+	}
+
+	/* memory for input/output buffers */
+	waves_codec->i_buffer = codec_allocate_memory(dev, waves_codec->buffer_bytes, 16);
+	if (!waves_codec->i_buffer) {
+		codec_free_memory(dev, waves_codec->response);
```

Comment:
please use goto's to clean up this exit logic

Reply:
done